File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofStores/AgentCreationStore.swift
import Foundation
import Observation
import WagerproofModels
import WagerproofServices

/// `AgentCreationStore` is the source of truth for the 6-step agent creation
/// wizard. Ports the `CreateAgentFormState` machinery from
/// `wagerproof-mobile/app/(drawer)/(tabs)/agents/create.tsx` (formState +
/// helpers) into a single observable.
///
/// State model:
///   - `draft` holds the in-progress agent — every field maps to the
///     RN `CreateAgentFormState` interface (`types/agent.ts:429-449`).
///   - `step` is the current wizard page (0..5).
///   - `archetypeRows` is the cached preset list — loaded once when the wizard
///     opens, mirroring RN's `usePresetArchetypes` query (30-min stale).
///   - `submitState` tracks the create-agent network call.
///
/// FIDELITY: Validation matches `validateScreen()` in RN line-for-line — the
/// view consults `canProceed(from:)` before allowing Next.
@Observable
@MainActor
public final class AgentCreationStore {
    public enum SubmitState: Equatable, Sendable {
        case idle
        case submitting
        case succeeded(Agent)
        case failed(String)
    }

    public enum ArchetypesLoadState: Equatable, Sendable {
        case idle
        case loading
        case loaded
        case failed(String)
    }

    /// Mutable draft snapshot. The wizard mutates fields directly via the
    /// store's setter helpers; views use `@Bindable` over the store to wire
    /// TextField / Toggle / Slider bindings.
    public struct Draft: Equatable, Sendable {
        public var preferredSports: [AgentSport] = []
        public var archetype: AgentArchetype? = nil
        public var name: String = ""
        // Defaults match RN INITIAL_FORM_STATE (types/agent.ts:451-462).
        public var avatarEmoji: String = "\u{1F916}" // 🤖
        public var avatarColor: String = "gradient:#6366f1,#ec4899"
        /// Chosen pixel character (avatar_0…avatar_7). nil = not chosen yet;
        /// hosts seed it once so the preview doesn't reshuffle per keystroke
        /// (create_agent has no sprite field — persisted via update_agent
        /// right after creation).
        public var spriteIndex: Int? = nil
        public var personalityParams: AgentPersonalityParams = .default
        public var customInsights: AgentCustomInsights = .empty
        public var autoGenerate: Bool = true
        public var autoGenerateTime: String = "09:00"
        public var autoGenerateTimezone: String = "America/New_York"

        /// Public no-arg init so callers outside the module (e.g. the
        /// onboarding agent-builder host) can construct a fresh draft
        /// when seeding from another source of truth.
        public init() {}
    }

    // MARK: - State

    public var draft = Draft()
    public var step: Int = 0
    public var archetypeRows: [PresetArchetypeRow] = []
    public var archetypesLoadState: ArchetypesLoadState = .idle
    public var submitState: SubmitState = .idle

    /// Names of agents the user already owns — set by the view after
    /// `AgentsStore.refresh()`. Used for duplicate-name validation, matching
    /// the RN check in `agents/create.tsx:167-178`.
    public var existingAgentNames: [String] = []

    public static let totalSteps = 6

    public init() {}

    // MARK: - Archetype loading

    /// Fetch the preset archetypes once. Idempotent — subsequent calls no-op
    /// if already loaded. The RN client uses React Query's 30-min staleTime
    /// for the equivalent caching.
    public func loadArchetypesIfNeeded() async {
        if case .loaded = archetypesLoadState { return }
        if case .loading = archetypesLoadState { return }
        archetypesLoadState = .loading
        do {
            let rows = try await PresetArchetypeService.fetchAll()
            self.archetypeRows = rows
            self.archetypesLoadState = .loaded
        } catch {
            self.archetypesLoadState = .failed((error as NSError).localizedDescription)
        }
    }

    // MARK: - Path selection & toggles

    /// Apply an archetype preset: auto-fills sports, archetype id, and merges
    /// the partial personality params on top of defaults. Mirrors RN's
    /// `applyArchetypePreset()` in create.tsx:141-159.
    public func applyArchetype(_ row: PresetArchetypeRow) {
        if let id = AgentArchetype(rawValue: row.id) {
            draft.archetype = id
        } else {
            // Row.id doesn't match a known enum case — leave archetype nil
            // (treated as fully-custom). The edge function accepts arbitrary
            // archetype strings server-side but we don't want a decode crash.
            draft.archetype = nil
        }
        draft.preferredSports = row.recommendedSports
        draft.personalityParams = .applying(row.personalityParams)
        draft.customInsights = row.customInsights
    }

    /// Reset to the "fully-custom" path: clears the archetype but keeps the
    /// user's chosen sports. Mirrors RN's `handleSelectPath('scratch')` flow.
    public func clearArchetype() {
        draft.archetype = nil
        draft.personalityParams = .default
        draft.customInsights = .empty
    }

    /// Toggle a sport selection. Plain multi-select — MLB used to be an
    /// exclusive pick (single-sport Statcast payload) but the generation
    /// pipeline now supports mixing it with other sports.
    public func toggleSport(_ sport: AgentSport) {
        let current = draft.preferredSports
        if current.contains(sport) {
            draft.preferredSports = current.filter { $0 != sport }
        } else {
            draft.preferredSports = current + [sport]
        }
    }

    // MARK: - Validation

    /// Whether the user can advance past the given step. Mirrors
    /// `validateScreen()` in create.tsx:162-192.
    public func canProceed(from stepIndex: Int) -> Bool {
        switch stepIndex {
        case 0:
            return !draft.preferredSports.isEmpty
        case 1:
            let trimmed = draft.name.trimmingCharacters(in: .whitespacesAndNewlines)
            if trimmed.isEmpty || trimmed.count > 50 { return false }
            let lowered = trimmed.lowercased()
            if existingAgentNames.contains(where: { $0.lowercased() == lowered }) { return false }
            if draft.avatarEmoji.isEmpty { return false }
            if draft.avatarColor.isEmpty { return false }
            return true
        case 2, 3, 4, 5:
            return true
        default:
            return false
        }
    }

    /// Surface-friendly explanation of why a step is blocked. Mirrors
    /// `getValidationError()` in create.tsx:195-226.
    public func validationError(for stepIndex: Int) -> String? {
        switch stepIndex {
        case 0:
            if draft.preferredSports.isEmpty { return "Please select at least one sport" }
            return nil
        case 1:
            let trimmed = draft.name.trimmingCharacters(in: .whitespacesAndNewlines)
            if trimmed.isEmpty { return "Please enter a name for your agent" }
            if trimmed.count > 50 { return "Name must be 50 characters or less" }
            let lowered = trimmed.lowercased()
            if existingAgentNames.contains(where: { $0.lowercased() == lowered }) {
                return "You already have an agent named \"\(trimmed)\". Please choose a different name."
            }
            if draft.avatarEmoji.isEmpty { return "Please select an emoji" }
            return nil
        default:
            return nil
        }
    }

    // MARK: - Navigation

    public func advance() {
        guard step < Self.totalSteps - 1 else { return }
        step += 1
    }

    public func back() {
        guard step > 0 else { return }
        step -= 1
    }

    // MARK: - Submission

    /// Submit the draft via the edge function. Sets `submitState` to either
    /// `.succeeded(agent)` or `.failed(reason)`.
    @discardableResult
    public func submit(autoModeForcedOff: Bool) async -> Agent? {
        let trimmedName = draft.name.trimmingCharacters(in: .whitespacesAndNewlines)
        // Honor the "Pro auto-slot full" gate the same way RN does — flip
        // auto_generate off before we serialize the payload so the server
        // sees the user's effective intent (manual mode).
        let shouldStartAuto = draft.autoGenerate && !autoModeForcedOff
        let input = CreateAgentInput(
            name: trimmedName,
            avatarEmoji: draft.avatarEmoji,
            avatarColor: draft.avatarColor,
            preferredSports: draft.preferredSports,
            archetype: draft.archetype,
            personalityParams: draft.personalityParams,
            customInsights: draft.customInsights,
            autoGenerate: shouldStartAuto,
            autoGenerateTime: draft.autoGenerateTime,
            autoGenerateTimezone: draft.autoGenerateTimezone
        )
        submitState = .submitting
        do {
            let agent = try await AgentService.create(input: input)
            submitState = .succeeded(agent)
            return agent
        } catch {
            submitState = .failed((error as NSError).localizedDescription)
            return nil
        }
    }
}

File: wagerproof-android-native/core/stores/src/main/kotlin/com/wagerproof/core/stores/AgentCreationStore.kt
package com.wagerproof.core.stores

import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.wagerproof.core.models.Agent
import com.wagerproof.core.models.AgentArchetype
import com.wagerproof.core.models.AgentCustomInsights
import com.wagerproof.core.models.AgentPersonalityParams
import com.wagerproof.core.models.AgentSport
import com.wagerproof.core.services.AgentService
import com.wagerproof.core.services.CreateAgentInput
import com.wagerproof.core.services.PresetArchetypeRow
import com.wagerproof.core.services.PresetArchetypeService
import com.wagerproof.core.services.applying
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel

/**
 * `AgentCreationStore` is the source of truth for the 6-step agent creation
 * wizard. Ports the `CreateAgentFormState` machinery from
 * `wagerproof-mobile/app/(drawer)/(tabs)/agents/create.tsx` (formState +
 * helpers) into a single observable.
 *
 * State model:
 *   - `draft` holds the in-progress agent — every field maps to the RN
 *     `CreateAgentFormState` interface (`types/agent.ts:429-449`).
 *   - `step` is the current wizard page (0..5).
 *   - `archetypeRows` is the cached preset list — loaded once when the wizard
 *     opens, mirroring RN's `usePresetArchetypes` query (30-min stale).
 *   - `submitState` tracks the create-agent network call.
 *
 * FIDELITY: Validation matches `validateScreen()` in RN line-for-line — the
 * view consults `canProceed(from:)` before allowing Next.
 */
@Stable
class AgentCreationStore {

    sealed interface SubmitState {
        data object Idle : SubmitState
        data object Submitting : SubmitState
        data class Succeeded(val agent: Agent) : SubmitState
        data class Failed(val message: String) : SubmitState
    }

    sealed interface ArchetypesLoadState {
        data object Idle : ArchetypesLoadState
        data object Loading : ArchetypesLoadState
        data object Loaded : ArchetypesLoadState
        data class Failed(val message: String) : ArchetypesLoadState
    }

    /**
     * Mutable draft snapshot. The wizard mutates fields via the store's setter
     * helpers (copy-based) so Compose recomposes on change.
     */
    data class Draft(
        val preferredSports: List<AgentSport> = emptyList(),
        val archetype: AgentArchetype? = null,
        val name: String = "",
        // Defaults match RN INITIAL_FORM_STATE (types/agent.ts:451-462).
        val avatarEmoji: String = "🤖", // 🤖
        val avatarColor: String = "gradient:#6366f1,#ec4899",
        /**
         * Chosen pixel character (avatar_0…avatar_7). null = not chosen yet;
         * hosts seed it once so the preview doesn't reshuffle per keystroke
         * (create_agent has no sprite field — persisted via update_agent right
         * after creation).
         */
        val spriteIndex: Int? = null,
        val personalityParams: AgentPersonalityParams = AgentPersonalityParams.default,
        val customInsights: AgentCustomInsights = AgentCustomInsights.empty,
        val autoGenerate: Boolean = true,
        val autoGenerateTime: String = "09:00",
        val autoGenerateTimezone: String = "America/New_York",
    )

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    // MARK: - State

    var draft by mutableStateOf(Draft())
    var step by mutableStateOf(0)
    var archetypeRows by mutableStateOf<List<PresetArchetypeRow>>(emptyList())
    var archetypesLoadState by mutableStateOf<ArchetypesLoadState>(ArchetypesLoadState.Idle)
    var submitState by mutableStateOf<SubmitState>(SubmitState.Idle)

    /**
     * Names of agents the user already owns — set by the view after
     * `AgentsStore.refresh()`. Used for duplicate-name validation.
     */
    var existingAgentNames by mutableStateOf<List<String>>(emptyList())

    // MARK: - Archetype loading

    /**
     * Fetch the preset archetypes once. Idempotent — subsequent calls no-op if
     * already loaded. RN uses React Query's 30-min staleTime for equivalent caching.
     */
    suspend fun loadArchetypesIfNeeded() {
        if (archetypesLoadState is ArchetypesLoadState.Loaded) return
        if (archetypesLoadState is ArchetypesLoadState.Loading) return
        archetypesLoadState = ArchetypesLoadState.Loading
        try {
            archetypeRows = PresetArchetypeService.fetchAll()
            archetypesLoadState = ArchetypesLoadState.Loaded
        } catch (t: Throwable) {
            archetypesLoadState = ArchetypesLoadState.Failed(message(t))
        }
    }

    // MARK: - Path selection & toggles

    /**
     * Apply an archetype preset: auto-fills sports, archetype id, and merges
     * the partial personality params on top of defaults. Mirrors RN's
     * `applyArchetypePreset()`.
     */
    fun applyArchetype(row: PresetArchetypeRow) {
        // Row.id may not match a known enum case (server can ship new ids) —
        // leave archetype null (fully-custom) rather than crashing on decode.
        val archetype = AgentArchetype.entries.firstOrNull { it.raw == row.id }
        draft = draft.copy(
            archetype = archetype,
            preferredSports = row.recommendedSports,
            personalityParams = AgentPersonalityParams.applying(row.personalityParams),
            customInsights = row.customInsights,
        )
    }

    /**
     * Reset to the "fully-custom" path: clears the archetype but keeps the
     * user's chosen sports. Mirrors RN's `handleSelectPath('scratch')` flow.
     */
    fun clearArchetype() {
        draft = draft.copy(
            archetype = null,
            personalityParams = AgentPersonalityParams.default,
            customInsights = AgentCustomInsights.empty,
        )
    }

    /**
     * Toggle a sport selection. Plain multi-select — MLB used to be exclusive
     * (single-sport Statcast payload) but the pipeline now supports mixing it.
     */
    fun toggleSport(sport: AgentSport) {
        val current = draft.preferredSports
        draft = if (current.contains(sport)) {
            draft.copy(preferredSports = current.filter { it != sport })
        } else {
            draft.copy(preferredSports = current + sport)
        }
    }

    // MARK: - Validation

    /** Whether the user can advance past the given step. Mirrors `validateScreen()`. */
    fun canProceed(from: Int): Boolean = when (from) {
        0 -> draft.preferredSports.isNotEmpty()
        1 -> {
            val trimmed = draft.name.trim()
            when {
                trimmed.isEmpty() || trimmed.length > 50 -> false
                existingAgentNames.any { it.lowercase() == trimmed.lowercase() } -> false
                draft.avatarEmoji.isEmpty() -> false
                draft.avatarColor.isEmpty() -> false
                else -> true
            }
        }
        2, 3, 4, 5 -> true
        else -> false
    }

    /** Surface-friendly explanation of why a step is blocked. Mirrors `getValidationError()`. */
    fun validationError(stepIndex: Int): String? = when (stepIndex) {
        0 -> if (draft.preferredSports.isEmpty()) "Please select at least one sport" else null
        1 -> {
            val trimmed = draft.name.trim()
            when {
                trimmed.isEmpty() -> "Please enter a name for your agent"
                trimmed.length > 50 -> "Name must be 50 characters or less"
                existingAgentNames.any { it.lowercase() == trimmed.lowercase() } ->
                    "You already have an agent named \"$trimmed\". Please choose a different name."
                draft.avatarEmoji.isEmpty() -> "Please select an emoji"
                else -> null
            }
        }
        else -> null
    }

    // MARK: - Navigation

    fun advance() {
        if (step < totalSteps - 1) step += 1
    }

    fun back() {
        if (step > 0) step -= 1
    }

    // MARK: - Submission

    /**
     * Submit the draft via the edge function. Sets `submitState` to either
     * `.succeeded(agent)` or `.failed(reason)`.
     */
    suspend fun submit(autoModeForcedOff: Boolean): Agent? {
        val trimmedName = draft.name.trim()
        // Honor the "Pro auto-slot full" gate — flip auto_generate off before
        // serializing so the server sees the user's effective intent (manual).
        val shouldStartAuto = draft.autoGenerate && !autoModeForcedOff
        val input = CreateAgentInput(
            name = trimmedName,
            avatarEmoji = draft.avatarEmoji,
            avatarColor = draft.avatarColor,
            preferredSports = draft.preferredSports,
            archetype = draft.archetype,
            personalityParams = draft.personalityParams,
            customInsights = draft.customInsights,
            autoGenerate = shouldStartAuto,
            autoGenerateTime = draft.autoGenerateTime,
            autoGenerateTimezone = draft.autoGenerateTimezone,
        )
        submitState = SubmitState.Submitting
        return try {
            val agent = AgentService.create(input)
            submitState = SubmitState.Succeeded(agent)
            agent
        } catch (t: Throwable) {
            submitState = SubmitState.Failed(message(t))
            null
        }
    }

    // MARK: - Lifecycle teardown

    fun close() = scope.cancel()

    companion object {
        const val totalSteps = 6

        private fun message(t: Throwable): String =
            t.message?.takeIf { it.isNotEmpty() } ?: "Unknown error"
    }
}
